libmultilabel/networks/bigru.py

```diff
@@ -0,0 +1,89 @@
+            is set to rnn_dim//2. Defaults to 512.
+        rnn_layers (int): Number of recurrent layers. Defaults to 1.
+        dropout (float): The dropout rate of the word embedding. Defaults to 0.2.
+        activation (str): Activation function to be used. Defaults to 'tanh'.
+    """
+    def __init__(
+        self,
+        embed_vecs,
+        num_classes,
+        rnn_dim=512,
+        rnn_layers=1,
+        dropout=0.2,
+        activation='tanh',
+        **kwargs
+    ):
+        super(BiGRU, self).__init__(embed_vecs, dropout, activation, **kwargs)
+        assert rnn_dim%2 == 0, """`rnn_dim` should be even."""
+
+        # BiGRU
+        emb_dim = embed_vecs.shape[1]
+        self.rnn = nn.GRU(emb_dim, rnn_dim//2, rnn_layers,
+                          bidirectional=True, batch_first=True)
+
+        # context vectors for computing attention
+        self.U = nn.Linear(rnn_dim, num_classes)
+        xavier_uniform_(self.U.weight)
+
+        # linear output
+        self.final = nn.Linear(rnn_dim, num_classes)
+        xavier_uniform_(self.final.weight)
+
+    def forward(self, input):
+        text, length, indices = self.sort_data_by_length(input['text'], input['length'])
+        x = self.embedding(text) # (batch_size, length, rnn_dim)
+        x = self.embed_drop(x) # (batch_size, length, rnn_dim)
+
+        packed_inputs = pack_padded_sequence(x, length, batch_first=True)
+        x, _ = self.rnn(packed_inputs)
+        x = pad_packed_sequence(x)[0]
+        x = x.permute(1, 0, 2)
+
+        x = torch.tanh(x)
+
+        # Apply per-label attention
+        alpha = torch.softmax(self.U.weight.matmul(x.transpose(1, 2)), dim=2)
+
+        # Document representations are weighted sums using the attention
+        m = alpha.matmul(x) # (batch_size, num_classes, rnn_dim)
+
+        # Compute a probability for each label
+        x = self.final.weight.mul(m).sum(dim=2).add(self.final.bias) # (batch_size, num_classes)
+
+        return {'logits': x[indices], 'attention': alpha[indices]}
+
+    def sort_data_by_length(self, text, length):
+        """Sort data by length. This function is the deterministic implementation of
+        `pack_padded_sequence` with `enforce_sorted = False`.
+
+        Args:
+            text (torch.Tensor): Batch of text input with shape (batch_size, length)
+            length (list): List of text lengths before padding.
+
+        Returns:
+            text (torch.Tensor): Text sorted by lengths in descending order.
+            length (torch.Tensor): Lengths sorted in descending order.
+            indices (torch.Tensor): The indexes of the elements in the original data.
+        """
+        length = torch.as_tensor(length, dtype=torch.int64)
+        length, sorted_indices = torch.sort(length, descending=True)
+        sorted_indices = sorted_indices.to(text.device)
+        text = text.index_select(0, sorted_indices)
+
+        _, indices = torch.sort(sorted_indices) # get the original order
```

Comment:
doing two sort isn't good. For the second sort, cannot we do some kind of index assignment. See the example below

>>> x = torch.randn(1,7)
>>> [z, ind] = torch.sort(x)
>>> y = torch.randn(1,7)
>>> ind
tensor([[3, 5, 6, 4, 1, 2, 0]])
>>> x
tensor([[ 1.7054,  0.3207,  1.5532, -2.0915,  0.2383, -0.4152,  0.0647]])
>>> y[0,3] = 0
>>> y[0,5] = 1
>>> y[0,6] = 2
>>> y[0,4] = 3
>>> y[0,1] = 4
>>> y[0,2] = 5
>>> y[0,0] = 6
>>> y
tensor([[6., 4., 5., 0., 3., 1., 2.]])
>>> 

though I don't know if we can do the above y assignment in just one line. I think it's possiblem

Reply:
Sure. This works!

```python
y = torch.clone(ind)
y.index_put_(tuple(ind.unsqueeze(0)), torch.arange(ind.size(-1)))
```